### PR TITLE
Bugfix for uninitialised position in PrefetchIterator

### PR DESCRIPTION
**Pull request by thomascorthals**
`PrefetchIterator` didn't have an initial value for its `$position` and that causes two bugs. You can test it by replacing the `foreach` loop in `examples/7.6-plugin-prefetchiterator.php`.

1) You got an error when calling `key()` on a fresh iterator if you didn't call `rewind()` first.

```php
for (; $prefetch->valid(); $prefetch->next()) {
    echo '<hr/>'.$prefetch->key().': '. $prefetch->current()->id;
}
```
```
Fatal error: Uncaught TypeError: Return value of Solarium\Plugin\PrefetchIterator::key() must be of the type int, null returned in solarium/src/Plugin/PrefetchIterator.php:205 Stack trace: #0 solarium/examples/7.6-plugin-prefetchiterator.php(26): Solarium\Plugin\PrefetchIterator->key() #1 {main} thrown in solarium/src/Plugin/PrefetchIterator.php on line 205
```

2) If you're doing a `count()` first and then loop through it without `rewind()`ing, the first set of fetched documents would be skipped. In our example, `$i` ended up being 30 for the 32 documents in techproducts.

```php
for ($i = 0; $prefetch->valid(); $prefetch->next(), ++$i) {
    echo '<hr/>'.$i.': '. $prefetch->current()->id;
}
```

These problems don't surface when you're using `foreach` to loop over the results because that does a `rewind()` first. But you would run into it with a `while` in a use case like this:
```php
while ($n < $max && $prefetch->valid()) {
    if (doCheck($prefetch->current()) {
        ++$n;
    }
    $prefetch->next();
}
```

Furthermore, since `count()` implements a method from `\Countable`, it shouldn't have any publically visible effects on the methods that are implemented from `\Iterator`. They're separate interfaces.

Both problems disappear when `$position` is initialised to `0`. In fact the [example in the PHP Manual](https://www.php.net/manual/en/class.iterator.php#iterator.examples) does just that.

I've added a manual walkthrough of a full iteration and some rewinds in the unit tests. I've also added tests to ensure an empty result set works as expected.

I do believe I must've run into the same issue when I first added the integration tests. I didn't include an `assertCount()` then and I think it was because it didn't agree with the loop. It's included now and the loop also checks if the documents are actually iterated in the correct order. This is a relevant test for #610.

### CODE REVIEW

File: src/Plugin/PrefetchIterator.php [hunk 1]
```diff
@@ -58,7 +58,7 @@ class PrefetchIterator extends AbstractPlugin implements \Iterator, \Countable
      *
      * @var int
      */
-    protected $position;
+    protected $position = 0;
```
Comment (wickedOne) [inline src/Plugin/PrefetchIterator.php:61]: isn't defining the default in the constructor sufficient?

Comment (thomascorthals) [reply src/Plugin/PrefetchIterator.php:61]: The example in the PHP Manual does both:
```php
<?php
class myIterator implements Iterator {
    private $position = 0;
    // ...  

    public function __construct() {
        $this->position = 0;
    }
```

As it is now, none of the plugin classes have an explicit constructor defined. The parent constructor — `Configurable::__construct()` — is called from `AbstractPlugin::__construct()`.

Following the logic of `Configurable`, this should be done in `PrefetchIterator::init()`? But then we might as well rename `resetData()` to `init()` …

Comment (wickedOne) [reply src/Plugin/PrefetchIterator.php:61]: ah yes, i see...

personally i think ``resetData`` is fine

**Review by wickedOne — APPROVED** [review]
